### PR TITLE
Copter: add support for Winch

### PR DESCRIPTION
**Pull request by rmackay9**
This PR adds support for a winch to lower packages from a vehicle:

- new AP_WINCH library added which allows controlling a winch for lower packages from a vehicle.
  - makes use of the AP_WheelEncoder library to monitor the length of line released and rate of release
  - allows controlling the winch either by specifying how much line should be wound or unwound or by specifying a rate of winding/unwinding.
- pilot can control the winch manually using:
  - auxiliary switch - enable/disable and/or raise/lower
  - channel 6 tuning knob (controls rate of release)
- AP_Mission has been enhanced to allow autonomous control of the winch either using rate or change in length.
- GCS can also issue a command-long or command-int to control the winch
- new DO_WINCH command has been added to mavlink (https://github.com/ArduPilot/mavlink/pull/54)

- this is failing travis because it requires this mavlink PR to be merged: https://github.com/ArduPilot/mavlink/pull/54

Hardware used for this was:

- pololu.com motor with encoder [like this one](https://www.pololu.com/product/2822)
- eLAB motor driver which is something like [this one](https://www.pololu.com/product/1451) but accepts regular pwm input (in the range 1000 ~ 2000).

Note:

- the mavlink module update to point to my repo should be replaced with the link to ardupilot/mavlink once the above mavlink related PR is merged

### CODE REVIEW

File: libraries/AP_Winch/AP_Winch.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+#pragma once
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_Param/AP_Param.h>
+#include <AC_PID/AC_PID.h>
+#include <AP_WheelEncoder/AP_WheelEncoder.h>
+
+// winch rate control default gains
+#define AP_WINCH_POS_P      1.00f   // hard coded position error to desired rate gain
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch.h]: Why not let these be user editable PID's (that only show up when the library is enabled). Seems like something that would need different tuning for different hardware.

File: libraries/AP_Winch/AP_Winch.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    typedef enum {
+        STATE_DISABLED = 0,             // winch is not operating
+        STATE_POSITION,                 // moving or maintaining a target length
+        STATE_RATE,                     // deploying or retracting at a target rate
+    }  winch_state;
+
+    struct Backend_Config {
+        AP_Int8     type;               // winch type
+        AP_Float    rate_max;           // deploy or retract rate maximum (in m/s).
+        AP_Float    length_curr;        // current length of the line (in meters) that has been deployed (parameter to allow pilot override)
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch.h]: This is not actually user overrideable, it simply gets blindly set with the wheel encoder distance in update() as far as I can tell. I would simply make this not a parameter as there isn't any handling of updates to it, nor can I find any notifications to assist a GCS with it.

File: libraries/AP_Winch/AP_Winch.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+void AP_Winch::release_length(float length)
+{
+    config.length_desired = config.length_curr + length;
+    config.state = STATE_POSITION;
+}
+
+// deploy line at specified speed in m/s (+ve deploys line, -ve retracts line, 0 stops)
+void AP_Winch::set_desired_rate(float rate)
+{
+    config.rate_desired = constrain_float(rate, -config.rate_max, config.rate_max);
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch.cpp]: This should use get_rate_max() because on any constrain if low > high then the constrain will fail to do anything, and if the user typed a negative rate max in it wouldn't be handled correctly. (An error condition for sure, but we don't stop them from putting in that kind of value in the param).

File: libraries/AP_Winch/AP_Winch.h (hunk 2 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.h]: Yes, it's not useful having the length_curr as a parameter.  I've removed it so it's just a regular internal variable now.

File: libraries/AP_Winch/AP_Winch.cpp (hunk 3 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.cpp]: sounds good, I'll change it to use get_rate_max()

File: libraries/AP_Winch/AP_Winch.h (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.h]: yes, you're right, I'll make this position-gain a parameter.

Thanks for the review!

**Comment by khancyr** [comment]
Randy hello, about the winch, how much did you test the distance calculation ? Because it isn't linear... rope width isn't even on the winch axe, rope got some elongation, the more you throw rope the less a turn on the winch free rope, etc. Last time I use a winch, I spend like a day to make diffenrent lenght mesurement and reach some fonctionnal lenght law with a linear regression

**Comment by WickedShell** [comment]
@khancyr the easiest way to measure the extended length (ignoring elongation) would be to mount the wheel encoder against where the wire feeds through (or on a deployment pulley) so that it was measuring the rope length after it was unspooled.

If you did that you could add a simple elongation factor. However what is proposed here is not absolute length control and the entire winch system is unaware of what the deployed length is, it is instead based upon a relative length deployment, which means you are unaware how much the rope is truly below you at any point.

@rmackay9 probably out of scope on this PR, but I hope you have limit switches to cut the motor if you over tighten/deploy. (Ideally supporting at least one GPIO to detect a limit switch when the end effector is completely retracted).

File: ArduCopter/GCS_Mavlink.cpp [hunk 4]
```diff
@@ -1194,6 +1194,34 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
             break;
 #endif
 
+        case MAV_CMD_DO_WINCH:
+            // param1 : winch number (ignored)
+            // param2 : action (0=release, 1=length control, 2=rate control). See WINCH_ACTIONS enum.
+            if(!copter.g2.winch.enabled()) {
+                result = MAV_RESULT_FAILED;
+            } else {
+                result = MAV_RESULT_ACCEPTED;
```
Comment (WickedShell) [inline ArduCopter/GCS_Mavlink.cpp]: This can be hit when the winch is enabled, but no backend is selected. This should only be accepted if a backend was actually created.

File: ArduCopter/GCS_Mavlink.cpp [hunk 5]
```diff
@@ -1194,6 +1194,34 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
...
+                switch ((uint8_t)packet.param2) {
+                    case WINCH_DISABLE:
+                        copter.g2.winch.relax();
+                        break;
+                    case WINCH_LENGTH_CONTROL: {
+                        float desired_length = copter.g2.winch.get_line_length() + packet.param3;
```
Comment (WickedShell) [inline ArduCopter/GCS_Mavlink.cpp]: Adding the current line length here is incorrect. You already do that inside ```AP_Winch::release_length()``` which is effectively increasing/decreasing the amount of length actually commanded. And inconsistent with how the command is used in a mission.

File: ArduCopter/GCS_Mavlink.cpp [hunk 6]
```diff
@@ -1194,6 +1194,34 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
...
+                        copter.g2.winch.release_length(desired_length);
+                        break;
+                    }
+                    case WINCH_RATE_CONTROL: {
+                        float desired_rate = packet.param4;
+                        copter.g2.winch.set_desired_rate(desired_rate);
```
Comment (WickedShell) [inline ArduCopter/GCS_Mavlink.cpp]: If a rate is requested that is outside the possible rate of the winch it should be a rejected command instead of simply clamping at the max rate. The problem with accepting it, is that it might convince a operator that more or less length has been deployed/retracted then actually has, and could cause the operator to plan a path/command a travel that would actually snag the line.

File: ArduCopter/commands_logic.cpp [hunk 7]
```diff
@@ -651,6 +655,29 @@ void Copter::do_guided_limits(const AP_Mission::Mission_Command& cmd)
...
 #endif
 
+// control winch based on mission command
+void Copter::do_winch(const AP_Mission::Mission_Command& cmd)
+{
+    // Note: we ignore the gripper num parameter because we only support one gripper
+    switch (cmd.content.winch.action) {
+        case WINCH_DISABLE:
+            g2.winch.relax();
+            Log_Write_Event(DATA_WINCH_DISABLE);
```
Comment (WickedShell) [inline ArduCopter/commands_logic.cpp]: Out of scope for this PR, so no need to address, but it would be nice if one day the Log_Write_Event stuff lived in Dataflash so that the winch library could automatically stick the events in.

File: ArduCopter/GCS_Mavlink.cpp [hunk 8]
```diff
@@ -1194,6 +1194,34 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
...
+                result = MAV_RESULT_FAILED;
+            } else {
+                result = MAV_RESULT_ACCEPTED;
+                switch ((uint8_t)packet.param2) {
+                    case WINCH_DISABLE:
+                        copter.g2.winch.relax();
+                        break;
+                    case WINCH_LENGTH_CONTROL: {
+                        float desired_length = copter.g2.winch.get_line_length() + packet.param3;
+                        copter.g2.winch.release_length(desired_length);
```
Comment (WickedShell) [inline ArduCopter/GCS_Mavlink.cpp]: How come there are not any Log_Write_Event's here for length/rate on the winch, but their are in the mission? (This is where the later out of scope comment would be nice)

File: libraries/AP_Winch/AP_Winch.h [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+class AP_Winch {
+public:
+    AP_Winch();
+
+    // indicate whether this module is enabled or not
+    bool enabled() const { return _enabled; }
+
+    // initialise the winch
+    void init(const AP_WheelEncoder *wheel_encoder = nullptr);
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch.h]: Is there a reason that the wheel encoder can't actually be a child object of the winch? As outlined it will be harder to have support for multiple winches or use other wheel encoders for future use. Why not create a wheel encoder for each winch backend here, and make it a part of the AP_Winch param table? (Unless wheel encoder has to many nested params to fit like that).

File: libraries/AP_Winch/AP_Winch_Servo.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    // return immediately if no wheel encoder
+    if (_wheel_encoder == nullptr) {
+        return;
+    }
+
+    // if not doing any control stop output pwm exit
+    if (config.state == AP_Winch::STATE_DISABLED) {
+        SRV_Channels::set_safety_limit(SRV_Channel::k_winch, SRV_Channel::SRV_CHANNEL_LIMIT_ZERO_PWM);
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch_Servo.cpp]: I might not be familiar enough with the SRV_Channels stuff, but isn't this a one time setup aspect that could be set in the `if (!_servo_out_init) {` block? It's also worth noting that this is running a ioctl every time you hit it, which is probably not what you want to be doing in flight.

File: libraries/AP_Winch/AP_Winch_Servo.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        rate_desired = position_error * config.pos_p;
+    }
+
+    // if doing rate control, set desired rate
+    if (config.state == AP_Winch::STATE_RATE) {
+        rate_desired = config.rate_desired;
+    }
+
+    // calculate rate error and pass to pid controller
+    float rate_error = rate_desired - rate;
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch_Servo.cpp]: Shouldn't rate desired always be clamped to be within the max winch rate? Otherwise a high power motor can actually throw the rope off the winch faster then the load can pull it down, and can lead to all sorts of tangling/wire problems below the vehicle.

File: ArduCopter/switches.cpp [hunk 12]
```diff
@@ -624,6 +625,38 @@ void Copter::do_aux_switch_function(int8_t ch_function, uint8_t ch_flag)
             }
 #endif
             break;
+
+        case AUXSW_WINCH_ENABLE:
+            switch (ch_flag) {
+                case AUX_SWITCH_HIGH:
+                    // high switch maintains current position
+                    g2.winch.release_length(0.0f);
+                    Log_Write_Event(DATA_WINCH_LENGTH_CONTROL);
```
Comment (OXINARF) [inline ArduCopter/switches.cpp]: I have nothing against writing these events, but the AUXSW_WINCH_CONTROL will override this and have no event written.

File: ArduCopter/switches.cpp [hunk 13]
```diff
@@ -624,6 +625,38 @@ void Copter::do_aux_switch_function(int8_t ch_function, uint8_t ch_flag)
...
+                    break;
+                default:
+                    // all other position disable winch
+                    g2.winch.relax();
+                    Log_Write_Event(DATA_WINCH_DISABLE);
```
Comment (OXINARF) [inline ArduCopter/switches.cpp]: What is the difference supposed to be between disable/relax and maintaining current position?

File: ArduCopter/commands_logic.cpp [hunk 14]
```diff
@@ -160,6 +160,10 @@ bool Copter::start_command(const AP_Mission::Mission_Command& cmd)
         break;
 #endif
 
+    case MAV_CMD_DO_WINCH:                             // Mission command to control winch
```
Comment (OXINARF) [inline ArduCopter/commands_logic.cpp]: Shouldn't we add this to the verify_command list? It will send a warning message if it isn't there.

File: ArduCopter/GCS_Mavlink.cpp (hunk 4 above)
Comment (rmackay9) [reply ArduCopter/GCS_Mavlink.cpp]: Txs.  enabled() now checks if the backend is created and _TYPE parameter is non-zero.

File: ArduCopter/GCS_Mavlink.cpp (hunk 5 above)
Comment (rmackay9) [reply ArduCopter/GCS_Mavlink.cpp]: great, GCS_MAVLink control fixed in follow-up commit.

File: ArduCopter/GCS_Mavlink.cpp (hunk 8 above)
Comment (rmackay9) [reply ArduCopter/GCS_Mavlink.cpp]: Txs, added log-write-event messages when winch is operated via MAVlink.

File: libraries/AP_Winch/AP_Winch.h (hunk 9 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.h]: @OXINARF has commented on this one.  It crossed my mind but would require some changes to the AP_WheelEncoder class and Rover to allow separating them out (a bit more like AC_PID objects).  I'm not sure which is better or more consistent with what we're doing.  In some ways our move towards singletons is a move away from what's being suggested.

File: libraries/AP_Winch/AP_Winch_Servo.cpp (hunk 10 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: yes, that was a typo, fixed in a follow up commit, thanks!

File: libraries/AP_Winch/AP_Winch_Servo.cpp (hunk 11 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: yes, it should be constrained, added!

File: ArduCopter/commands_logic.cpp (hunk 14 above)
Comment (rmackay9) [reply ArduCopter/commands_logic.cpp]: added MAV_CMD_DO_WINCH to the verify_command functions too, thanks.

File: ArduCopter/switches.cpp (hunk 12 above)
Comment (rmackay9) [reply ArduCopter/switches.cpp]: I don't quite understand why these events wouldn't be written..

File: ArduCopter/switches.cpp (hunk 13 above)
Comment (rmackay9) [reply ArduCopter/switches.cpp]: "relax" will allow the motor to move while maintaining the current position will cause the controller to increase/decrease the motor output to maintain the rope at the current position.

File: libraries/AP_Mission/AP_Mission.cpp [hunk 15]
```diff
@@ -805,6 +805,13 @@ MAV_MISSION_RESULT AP_Mission::mavlink_int_to_mission_cmd(const mavlink_mission_
         cmd.content.set_yaw_speed.relative_angle = packet.param3;   // 0 = absolute angle, 1 = relative angle
         break;
 
+    case MAV_CMD_DO_WINCH:                              // MAV ID: 42600
+        cmd.content.winch.num = packet.param1;          // winch number
+        cmd.content.winch.action = packet.param2;       // action (0 = relax, 1 = length control, 2 = rate control).  See WINCH_ACTION enum
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Mission/AP_Mission.cpp]: Yup, ok.  I've renamed everything to "Relaxed" in a follow up commit.

File: libraries/AP_Winch/AP_Winch.cpp [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    // @Description: Winch enable/disable
+    // @User: Standard
+    // @Values: 0:Disabled, 1:Enabled
+    AP_GROUPINFO_FLAGS("_ENABLE", 0, AP_Winch, _enabled, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: TYPE
+    // @DisplayName: Winch Type
+    // @Description: Winch enable/disable
+    // @User: Standard
+    // @Values: 0:None,1:Servo with encoder
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.cpp]: That's a good point about not needing the 0:None option and setting the default to "1".  I've done this in a follow-up commit.  I somehow feel we should reserve the type of zero to mean none but I don't have a very specific reason.

File: libraries/AP_Winch/AP_Winch.h [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+    // deploy line at specified speed in m/s (+ve deploys line, -ve retracts line, 0 stops)
+    void set_desired_rate(float rate);
+
+    // get rate maximum in m/s
+    float get_rate_max() const { return MAX(config.rate_max, 0.0f); }
+
+    static const struct AP_Param::GroupInfo        var_info[];
+
+    // parameters
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.h]: made everything below, private, thanks.

File: libraries/AP_Winch/AP_Winch.h [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+    AP_Int8     _enabled;               // grabber enable/disable
+
+    // winch states
+    typedef enum {
+        STATE_DISABLED = 0,             // winch is not operating
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.h]: yup, renamed to relaxed.

File: libraries/AP_Winch/AP_Winch.h [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        STATE_POSITION,                 // moving or maintaining a target length
+        STATE_RATE,                     // deploying or retracting at a target rate
+    }  winch_state;
+
+    struct Backend_Config {
+        AP_Int8     type;               // winch type
+        AP_Float    rate_max;           // deploy or retract rate maximum (in m/s).
+        AP_Float    pos_p;              // position error P gain
+        AC_PID      rate_pid = AC_PID(AP_WINCH_RATE_P, AP_WINCH_RATE_I, AP_WINCH_RATE_D, AP_WINCH_RATE_IMAX, AP_WINCH_RATE_FILT, AP_WINCH_RATE_DT);           // rate control PID
+        winch_state state;              // state of winch control (using target position or targt rate)
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.h]: typo fixed, thanks.

File: libraries/AP_Winch/AP_Winch.h [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+        float       length_curr;        // current length of the line (in meters) that has been deployed (parameter to allow pilot override)
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.h]: comment fixed, txs

File: libraries/AP_Winch/AP_Winch_Servo.cpp [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+{
+    // return immediately if no servo is assigned to control the winch
+    if (!SRV_Channels::function_assigned(SRV_Channel::k_winch)) {
+        return;
+    }
+
+    if (!_servo_out_init) {
+        // set servo output range
+        SRV_Channels::set_angle(SRV_Channel::k_winch,  1000);
+        _servo_out_init = true;
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: moved to init.  I was thinking that it's possible the user might not have set-up the SERVOx_FUNCTION at first and initialising it within the update would allow us to init the range when they do set it up.  Still, it's simpler to do it within the init (saves a variable) and the original method wasn't fool-proof anyway (i.e. if the user moved the channel, the init wouldn't happen).

File: libraries/AP_Winch/AP_Winch_Servo.cpp [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+
+    // if doing position control, calculate position error to desired rate
+    float rate_desired = 0.0f;
+    if (config.state == AP_Winch::STATE_POSITION) {
+        float position_error = config.length_desired - config.length_curr;
+        rate_desired = position_error * config.pos_p;
+    }
+
+    // if doing rate control, set desired rate
+    if (config.state == AP_Winch::STATE_RATE) {
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: Changed set_safety_limit to set_output_limit, thanks for catching that bug.
I don't think we need an "else if" in the AP_Winch_Servo.cpp after the "config.state == AP_Winch::STATE_RATE" check.  The way it works is if it's doing position control, it sets the rate_desired based on the position-error and then falls through to the rate control.  If it's doing rate control then it skips over the position control part and just uses the rate-desired set by the switch or mavlink command.

**Comment by rmackay9** [comment]
I think I've addressed most of the critical issues raised above.
Re @khancyr's and @WickedShell's valid questions about the accuracy of the length calculations and adding the ability to detect and stop the winch before it hits it's limits, there's are good valid points but I don't think we should hold up this PR going into master to address these enhancements.  What's here adds value I think and these points can be addressed in follow-up commits if others are interested in adding them.

**Comment by WickedShell** [comment]
@rmackay9 I think I'd object that at a minimum this needs to be mad clear at a interface level (ie the mavlink commands) that this is a relative distance traveled, and that there is a future way to have an absolute length control without breaking everything a current user would be familiar with.

**Comment by rmackay9** [comment]
@WickedShell, can you be more specific about what change would help?  Is it renaming the WINCH_ACTIONS enum to be "WINCH_RELATIVE_LENGTH_CONTROL" instead of "WINCH_LENGTH_CONTROL"?

**Comment by WickedShell** [comment]
Relative length would be better, but it starts to hit a combinitorix problem as in the future to support absolute distance will double the number of things we have to handle. I have to wonder if adding some other unused param as relative vs absolute flag? (I can be convinced either way)

File: ArduCopter/switches.cpp (hunk 12 above)
Comment (OXINARF) [reply ArduCopter/switches.cpp]: These will be written. What I'm saying is that in AUXSW_WINCH_CONTROL it doesn't write any event. So even when an user uses the rate control with an aux switch, the log can still show length control as the latest event.

File: libraries/AP_Winch/AP_Winch.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/AP_Winch/AP_Winch.cpp]: Yes, I can see that None would be useful in case of multiple winch instances, so I'm OK with leaving that reserved for the future. I just think that with the single instance there is in this PR, it isn't worth it having the user have to set ENABLE and TYPE.

File: libraries/AP_Winch/AP_Winch_Servo.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: Yes, not needed. It just avoids an additional check when doing position control 🙂

File: ArduCopter/GCS_Mavlink.cpp [hunk 23]
```diff
@@ -1199,6 +1199,39 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
             break;
 #endif
 
+        case MAV_CMD_DO_WINCH:
+            // param1 : winch number (ignored)
+            // param2 : action (0=disable, 1=length control, 2=rate control). See WINCH_ACTIONS enum.
```
Comment (OXINARF) [inline ArduCopter/GCS_Mavlink.cpp]: disable -> relax

File: ArduCopter/tuning.cpp [hunk 24]
```diff
@@ -212,5 +212,17 @@ void Copter::tuning() {
...
+
+     case TUNING_WINCH: {
+         float desired_rate = 0.0f;
+         if (v > 0.8f) {
+             desired_rate = g2.winch.get_rate_max();
+         }
+         if (v < 0.2f) {
+             desired_rate = -g2.winch.get_rate_max();
+         }
+         g2.winch.set_desired_rate(desired_rate);
```
Comment (OXINARF) [inline ArduCopter/tuning.cpp]: Isn't this supposed to be a linear change? Unless I'm missing something this is doing the same as the switch.

File: ArduCopter/GCS_Mavlink.cpp [hunk 25]
```diff
@@ -1194,6 +1194,34 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
             break;
 #endif
 
+        case MAV_CMD_DO_WINCH:
+            // param1 : winch number (ignored)
+            // param2 : action (0=release, 1=length control, 2=rate control). See WINCH_ACTIONS enum.
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply ArduCopter/GCS_Mavlink.cpp]: yup, changed to relaxed in the latest commits

File: ArduCopter/switches.cpp (hunk 12 above)
Comment (rmackay9) [reply ArduCopter/switches.cpp]: OK, I think I understand.  So the issue is that if the ch6 tuning knob is used (i.e see tuning.cpp) then there are no DATA_WINCH_RATE_CONTROL messages written to the log files and also the ch6 control will conflict with the aux switch controls if they are setup (the winch would momentarily try to do length control if the aux switches were changed but would then immediately revert to rate control).  I think we can handle this with some documentation on the still-to-be-created wiki page for the winch. I don't think the missing event messages for the rate-control while using ch6 tuning are too important.

File: libraries/AP_Winch/AP_Winch.cpp [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+    // @Param: ENABLE
+    // @DisplayName: Winch enable/disable
+    // @Description: Winch enable/disable
+    // @User: Standard
+    // @Values: 0:Disabled, 1:Enabled
+    AP_GROUPINFO_FLAGS("_ENABLE", 0, AP_Winch, _enabled, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: TYPE
+    // @DisplayName: Winch Type
+    // @Description: Winch enable/disable
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.cpp]: fixed.

File: libraries/AP_Winch/AP_Winch.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.cpp]: Yes, I agree it's slightly annoying to have both enable and type.  I did think of combining them but I also want to hide the parameters because 99.9% of users will never use a winch.  I thought hiding the other parameters with the winch-type parameter might be odd.. at least people might not be expecting a parameter ending in "_TYPE" to have the same parameter hiding effect as we sometimes do with "_ENABLE" parameters.

**Comment by rmackay9** [comment]
OK, I've made some more changes, hopefully this is it for now.

File: ArduCopter/GCS_Mavlink.cpp [hunk 27]
```diff
@@ -1199,6 +1199,39 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
...
+                        copter.g2.winch.relax();
+                        copter.Log_Write_Event(DATA_WINCH_RELAXED);
+                        break;
+                    case WINCH_RELATIVE_LENGTH_CONTROL: {
+                        copter.g2.winch.release_length(packet.param3, fabsf(packet.param4));
+                        copter.Log_Write_Event(DATA_WINCH_LENGTH_CONTROL);
+                        break;
+                    }
+                    case WINCH_RATE_CONTROL: {
+                        if (fabsf(packet.param4) < copter.g2.winch.get_rate_max()) {
```
Comment (WickedShell) [inline ArduCopter/GCS_Mavlink.cpp]: This should be <= otherwise you can never use your maximum deployment rate.

File: libraries/AP_Mission/AP_Mission.h [hunk 28]
```diff
@@ -184,6 +184,14 @@ class AP_Mission {
         uint8_t relative_angle; // 0 = absolute angle, 1 = relative angle
     };
 
+    // winch command structure
+    struct PACKED Winch_Command {
```
Comment (WickedShell) [inline libraries/AP_Mission/AP_Mission.h]: It is not an actual/blocking issue at all, but this uses up all the space that is available for the command in storage except for the 2 byte general storage parameter that all commands get as p1. Off the top of my head I can't come up with any other fields for a winch though that we might want in the future other then end effector controls, which really are best done as their own commands anyways, so I think this is fine.

File: libraries/AP_Winch/AP_Winch_Servo.cpp [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+    // calculate latest rate
+    float distance = _wheel_encoder->get_distance(0);
+    float rate = 0.0f;
+    if (is_positive(dt)) {
+        rate = (distance - _last_distance) / dt;
+    }
+    _last_distance = distance;
+
+    // update distance from wheel encoder
+    config.length_curr = distance;
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch_Servo.cpp]: `_last_distance` and `config.length_curr` always track the same value, can `_last_distance` be dropped in favour of the other?

File: libraries/AP_Winch/AP_Winch.cpp [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+        config.rate_desired = config.rate_max;
+    } else {
+        config.rate_desired = constrain_float(fabsf(rate), -config.rate_max, config.rate_max);
+    }
+}
+
+// deploy line at specified speed in m/s (+ve deploys line, -ve retracts line, 0 stops)
+void AP_Winch::set_desired_rate(float rate)
+{
+    config.rate_desired = constrain_float(rate, -get_rate_max(), get_rate_max());
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch.cpp]: Why is this fetching the maximum rate differently then how release length is fetching it?

File: libraries/AP_Winch/AP_Winch_Servo.cpp [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+    // calculate rate error and pass to pid controller
+    float rate_error = rate_desired - rate;
+    config.rate_pid.set_input_filter_all(rate_error);
+
+    // get p
+    float p = config.rate_pid.get_p();
+
+    // get i unless winch hit limit on last iteration
+    float i = config.rate_pid.get_integrator();
+    if (((is_negative(rate_error) && !limit_low) || (is_positive(rate_error) && !limit_high))) {
```
Comment (WickedShell) [inline libraries/AP_Winch/AP_Winch_Servo.cpp]: I think I'm missing something here, but if on a loop through (with lets say a large integrated value, but a fairly low I grain, we could hit the output limit on one round because the integrator has built up high, so we output a max value value. On the next loop through the limit_high will be set, so we will instead put out the I gain. This is potentially small enough to drop us below the maximum output so we will only output a less then max value this update and unset the flag. The next update however we will go back to maxing it out.

Basically I think I'm seeing a possible control induced pulsing of the winch when operating near limits. I could easily be missing something about how the underlying controller is actually behaving though.

File: ArduCopter/GCS_Mavlink.cpp (hunk 27 above)
Comment (rmackay9) [reply ArduCopter/GCS_Mavlink.cpp]: great, thanks!

File: libraries/AP_Winch/AP_Winch.cpp (hunk 30 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch.cpp]: no good reason, I'll change release_length to use get-rate-max.

File: libraries/AP_Winch/AP_Winch_Servo.cpp (hunk 29 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: nice optimisation.  dropping _last_distance.

File: libraries/AP_Winch/AP_Winch_Servo.cpp (hunk 31 above)
Comment (rmackay9) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: In case it's not clear the "I gain" (i.e. "config.rate_pid.get_integrator()") will be the integrator value from the previous iteration in which the output hit the limit.  So the final output would only drop if the actual error has dropped.

Hopefully the integrator output won't be changing that quickly so hopefully the result of this is that the integrator stops building up at moments when the output is hitting the limits.

We do something very similar in Copter's attitude control.  I've tried to explain this on page 6 of this power point: https://www.dropbox.com/s/wleq7imwtgo468d/Randy_Elab_Offsite_Aug2017.pptx?dl=0

Comment (WickedShell) [reply libraries/AP_Winch/AP_Winch_Servo.cpp]: Ah, okay I thought "I gain" was the actual tuning parameter of the I term, not the previous I value :)
